File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Barbados.m
//
//  UIImage+Barbados.m
//  CountriesFlags
//

#import "Countries/UIImage+Barbados.h"

@implementation UIImage (Barbados)

+ (UIImage *)flagBarbadosWithSize:(CGSize)size {
    UIColor* colorBlue = [UIColor colorWithRed:1.0f/255.0f
                                         green:38.0f/255.0f
                                          blue:127.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorYellow = [UIColor colorWithRed:1.0f
                                           green:199.0f/255.0f
                                            blue:38.0f/255.0f
                                           alpha:1.0f];
    UIColor* colorBlack = [UIColor blackColor];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorBlue setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorYellow setFill];
    UIRectFill(CGRectMake(size.width / 3.0f, 0.0f, size.width / 3.0f, size.height));
    
    [colorBlack setFill];
    UIBezierPath* path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointMake(size.width / 100.0 * 48.2f, size.height / 100.0f * 74.1f)];
    [path addLineToPoint:CGPointMake(size.width / 100.0f * 48.2f, size.height / 100.0f * 60.9f)];
    // left side of trident
    [path addCurveToPoint:CGPointMake(size.width / 100.0f * 42.1f, size.height / 100.0f * 62.2f)
            controlPoint1:CGPointMake(size.width / 100.0f * 47.2f, size.height / 100.0f * 60.9f)
            controlPoint2:CGPointMake(size.width / 100.0f * 43.1f, size.height / 100.0f * 62.2f)];
    [path addCurveToPoint:CGPointMake(size.width / 100.0f * 35.6f, size.height / 100.0f * 32.9f)
            controlPoint1:CGPointMake(size.width / 100.0f * 42.1f, size.height / 100.0f * 50.0f)
            controlPoint2:CGPointMake(size.width / 100.0f * 39.0f, size.height / 100.0f * 41.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 100.0f * 43.1f, size.height / 100.0f * 38.3f)
            controlPoint1:CGPointMake(size.width / 100.0f * 35.6f, size.height / 100.0f * 32.2f)
            controlPoint2:CGPointMake(size.width / 100.0f * 40.0f, size.height / 100.0f * 33.6f)];
    [path addCurveToPoint:CGPointMake(size.width / 100.0f * 41.5f, size.height / 100.0f * 38.4f)
            controlPoint1:CGPointMake(size.width / 100.0f * 42.3f, size.height / 100.0f * 38.25f)
            controlPoint2:CGPointMake(size.width / 100.0f * 41.5f, size.height / 100.0f * 37.3f)];
    [path addCurveToPoint:CGPointMake(size.width / 100.0f * 44.65f, size.height / 100.0f * 57.1f)
            controlPoint1:CGPointMake(size.width / 100.0f * 41.8f, size.height / 100.0f * 39.0f)
            controlPoint2:CGPointMake(size.width / 100.0f * 44.6f, size.height / 100.0f * 50.0f)];
    
    // central part of trident
    [path addLineToPoint:CGPointMake(size.width / 100.0f * 48.2f, size.height / 100.0f * 56.44f)];
    [path addLineToPoint:CGPointMake(size.width / 100.0f * 48.2f, size.height / 100.0f * 37.7f)];
    [path addCurveToPoint:CGPointMake(size.width / 100.0f * 45.3f, size.height / 100.0f * 38.5f)
            controlPoint1:CGPointMake(size.width / 100.0f * 47.5f, size.height / 100.0f * 37.7f)
            controlPoint2:CGPointMake(size.width / 100.0f * 46.0f, size.height / 100.0f * 38.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 100.0f * 49.9f, size.height / 100.0f * 25.7f)
            controlPoint1:CGPointMake(size.width / 100.0f * 47.9f, size.height / 100.0f * 35.0f)
            controlPoint2:CGPointMake(size.width / 100.0f * 49.7f, size.height / 100.0f * 27.5f)];
    [path addCurveToPoint:CGPointMake(size.width / 100.0f * 54.5f, size.height / 100.0f * 38.5f)
            controlPoint1:CGPointMake(size.width / 100.0f * 51.5f, size.height / 100.0f * 32.0f)
            controlPoint2:CGPointMake(size.width / 100.0f * 52.5f, size.height / 100.0f * 35.1f)];
    [path addCurveToPoint:CGPointMake(size.width / 100.0f * 51.6f, size.height / 100.0f * 37.7f)
            controlPoint1:CGPointMake(size.width / 100.0f * 54.0f, size.height / 100.0f * 38.0f)
            controlPoint2:CGPointMake(size.width / 100.0f * 52.1f, size.height / 100.0f * 37.7f)];
    [path addLineToPoint:CGPointMake(size.width / 100.0f * 51.6f, size.height / 100.0f * 56.44f)];
    [path addLineToPoint:CGPointMake(size.width / 100.0f * 55.15f, size.height / 100.0f * 57.1f)];

    // right side of trident
    [path addCurveToPoint:CGPointMake(size.width / 100.0f * 58.0f, size.height / 100.0f * 39.0f)
            controlPoint1:CGPointMake(size.width / 100.0f * 55.1f, size.height / 100.0f * 53.0f)
            controlPoint2:CGPointMake(size.width / 100.0f * 57.1f, size.height / 100.0f * 42.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 100.0f * 56.7f, size.height / 100.0f * 38.3f)
            controlPoint1:CGPointMake(size.width / 100.0f * 58.5f, size.height / 100.0f * 37.0f)
            controlPoint2:CGPointMake(size.width / 100.0f * 57.5f, size.height / 100.0f * 38.1f)];
    [path addCurveToPoint:CGPointMake(size.width / 100.0f * 64.2f, size.height / 100.0f * 32.9f)
            controlPoint1:CGPointMake(size.width / 100.0f * 58.0f, size.height / 100.0f * 35.0f)
            controlPoint2:CGPointMake(size.width / 100.0f * 64.0f, size.height / 100.0f * 32.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 100.0f * 57.7f, size.height / 100.0f * 62.2f)
            controlPoint1:CGPointMake(size.width / 100.0f * 60.9f, size.height / 100.0f * 41.0f)
            controlPoint2:CGPointMake(size.width / 100.0f * 57.7f, size.height / 100.0f * 50.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 100.0f * 51.6f, size.height / 100.0f * 60.9f)
            controlPoint1:CGPointMake(size.width / 100.0f * 56.7f, size.height / 100.0f * 62.2f)
            controlPoint2:CGPointMake(size.width / 100.0f * 52.6f, size.height / 100.0f * 60.9f)];
    [path addLineToPoint:CGPointMake(size.width / 100.0f * 51.6f, size.height / 100.0f * 74.1f)];
    [path closePath];
    [path fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image != nil ? image : [UIImage new];
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Barbados.swift
//
//  UIImage+Barbados.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Barbados "wikipedia"
     
     Generates the flag of Barbados.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Barbados.
     */
    internal static func flagBarbados(with size: CGSize) -> UIImage {
        let colorBlue = UIColor(red: 1.0 / 255.0,
                                green: 38.0 / 255.0,
                                blue: 127.0 / 255.0,
                                alpha: 1.0)
        let colorYellow = UIColor(red: 1.0,
                                  green: 199.0 / 255.0,
                                  blue: 38.0 / 255.0,
                                  alpha: 1.0)
        let colorBlack = UIColor.black

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorBlue.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorYellow.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: size.width / 3.0, y: 0.0),
                          size: CGSize(width: size.width / 3.0, height: size.height)))

        colorBlack.setFill()
        let path = UIBezierPath()
        path.move(to: CGPoint(x: size.width / 100.0 * 48.2, y: size.height / 100.0 * 74.1));
        path.addLine(to: CGPoint(x: size.width / 100.0 * 48.2, y: size.height / 100.0 * 60.9));
        // left side of trident
        path.addCurve(to: CGPoint(x: size.width / 100.0 * 42.1, y: size.height / 100.0 * 62.2),
                      controlPoint1: CGPoint(x: size.width / 100.0 * 47.2, y: size.height / 100.0 * 60.9),
                      controlPoint2: CGPoint(x: size.width / 100.0 * 43.1, y: size.height / 100.0 * 62.2))
        path.addCurve(to: CGPoint(x: size.width / 100.0 * 35.6, y: size.height / 100.0 * 32.9),
                      controlPoint1: CGPoint(x: size.width / 100.0 * 42.1, y: size.height / 100.0 * 50.0),
                      controlPoint2: CGPoint(x: size.width / 100.0 * 39.0, y: size.height / 100.0 * 41.0))
        path.addCurve(to: CGPoint(x: size.width / 100.0 * 43.1, y: size.height / 100.0 * 38.3),
                      controlPoint1: CGPoint(x: size.width / 100.0 * 35.6, y: size.height / 100.0 * 32.2),
                      controlPoint2: CGPoint(x: size.width / 100.0 * 40.0, y: size.height / 100.0 * 33.6))
        path.addCurve(to: CGPoint(x: size.width / 100.0 * 41.5, y: size.height / 100.0 * 38.4),
                      controlPoint1: CGPoint(x: size.width / 100.0 * 42.3, y: size.height / 100.0 * 38.25),
                      controlPoint2: CGPoint(x: size.width / 100.0 * 41.5, y: size.height / 100.0 * 37.3))
        path.addCurve(to: CGPoint(x: size.width / 100.0 * 44.65, y: size.height / 100.0 * 57.1),
                      controlPoint1: CGPoint(x: size.width / 100.0 * 41.8, y: size.height / 100.0 * 39.0),
                      controlPoint2: CGPoint(x: size.width / 100.0 * 44.6, y: size.height / 100.0 * 50.0))
        
        // central part of trident
        path.addLine(to: CGPoint(x: size.width / 100.0 * 48.2, y: size.height / 100.0 * 56.44));
        path.addLine(to: CGPoint(x: size.width / 100.0 * 48.2, y: size.height / 100.0 * 37.7));
        path.addCurve(to: CGPoint(x: size.width / 100.0 * 45.3, y: size.height / 100.0 * 38.5),
                      controlPoint1: CGPoint(x: size.width / 100.0 * 47.5, y: size.height / 100.0 * 37.7),
                      controlPoint2: CGPoint(x: size.width / 100.0 * 46.0, y: size.height / 100.0 * 38.0))
        path.addCurve(to: CGPoint(x: size.width / 100.0 * 49.9, y: size.height / 100.0 * 25.7),
                      controlPoint1: CGPoint(x: size.width / 100.0 * 47.9, y: size.height / 100.0 * 35.0),
                      controlPoint2: CGPoint(x: size.width / 100.0 * 49.7, y: size.height / 100.0 * 27.5))
        path.addCurve(to: CGPoint(x: size.width / 100.0 * 54.5, y: size.height / 100.0 * 38.5),
                      controlPoint1: CGPoint(x: size.width / 100.0 * 51.5, y: size.height / 100.0 * 32.0),
                      controlPoint2: CGPoint(x: size.width / 100.0 * 52.5, y: size.height / 100.0 * 35.1))
        path.addCurve(to: CGPoint(x: size.width / 100.0 * 51.6, y: size.height / 100.0 * 37.7),
                      controlPoint1: CGPoint(x: size.width / 100.0 * 54.0, y: size.height / 100.0 * 38.0),
                      controlPoint2: CGPoint(x: size.width / 100.0 * 52.1, y: size.height / 100.0 * 37.7))
        path.addLine(to: CGPoint(x: size.width / 100.0 * 51.6, y: size.height / 100.0 * 56.44));
        path.addLine(to: CGPoint(x: size.width / 100.0 * 55.15, y: size.height / 100.0 * 57.1));

        // right side of trident
        path.addCurve(to: CGPoint(x: size.width / 100.0 * 58.0, y: size.height / 100.0 * 39.0),
                      controlPoint1: CGPoint(x: size.width / 100.0 * 55.1, y: size.height / 100.0 * 53.0),
                      controlPoint2: CGPoint(x: size.width / 100.0 * 57.1, y: size.height / 100.0 * 42.0))
        path.addCurve(to: CGPoint(x: size.width / 100.0 * 56.7, y: size.height / 100.0 * 38.3),
                      controlPoint1: CGPoint(x: size.width / 100.0 * 58.5, y: size.height / 100.0 * 37.0),
                      controlPoint2: CGPoint(x: size.width / 100.0 * 57.5, y: size.height / 100.0 * 38.1))
        path.addCurve(to: CGPoint(x: size.width / 100.0 * 64.2, y: size.height / 100.0 * 32.9),
                      controlPoint1: CGPoint(x: size.width / 100.0 * 58.0, y: size.height / 100.0 * 35.0),
                      controlPoint2: CGPoint(x: size.width / 100.0 * 64.0, y: size.height / 100.0 * 32.0))
        path.addCurve(to: CGPoint(x: size.width / 100.0 * 57.7, y: size.height / 100.0 * 62.2),
                      controlPoint1: CGPoint(x: size.width / 100.0 * 60.9, y: size.height / 100.0 * 41.0),
                      controlPoint2: CGPoint(x: size.width / 100.0 * 57.7, y: size.height / 100.0 * 50.0))
        path.addCurve(to: CGPoint(x: size.width / 100.0 * 51.6, y: size.height / 100.0 * 60.9),
                      controlPoint1: CGPoint(x: size.width / 100.0 * 56.7, y: size.height / 100.0 * 62.2),
                      controlPoint2: CGPoint(x: size.width / 100.0 * 52.6, y: size.height / 100.0 * 60.9))
        path.addLine(to: CGPoint(x: size.width / 100.0 * 51.6, y: size.height / 100.0 * 74.1));
        path.close()
        path.fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image ?? UIImage()
    }
}

#endif
